### PR TITLE
change cert generate method and add pd and kv webhook

### PR DESCRIPTION
**Pull request by shuijing198799**
1. auto generator ca key crt file
2. add kv and pd webhook

### CODE REVIEW

**Comment by shuijing198799** [comment]
@weekface @xiaojingchen PTLA

**Comment by xiaojingchen** [comment]
@onlymellb PTAL the cert generate method

File: tests/pkg/apimachinery/certs.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	Key         []byte
+	SigningCert []byte
+}
+
+// Setup the server cert. For example, user apiservers and admission webhooks
+// can use the cert to prove their identify to the kube-apiserver
+func SetupServerCert(namespaceName, serviceName string) *CertContext {
+	certDir, err := ioutil.TempDir("", "test-e2e-server-cert")
+	if err != nil {
+		glog.Errorf("Failed to create a temp dir for cert generation %v", err)
```
Comment (weekface) [inline tests/pkg/apimachinery/certs.go]: You should return an error.

File: tests/pkg/apimachinery/certs.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	defer os.RemoveAll(certDir)
+	signingKey, err := cert.NewPrivateKey()
+	if err != nil {
+		glog.Errorf("Failed to create CA private key %v", err)
+	}
+	signingCert, err := cert.NewSelfSignedCACert(cert.Config{CommonName: "e2e-server-cert-ca"}, signingKey)
+	if err != nil {
+		glog.Errorf("Failed to create CA cert for apiserver %v", err)
```
Comment (weekface) [inline tests/pkg/apimachinery/certs.go]: ditto

File: tests/pkg/apimachinery/certs.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	caCertFile, err := ioutil.TempFile(certDir, "ca.crt")
+	if err != nil {
+		glog.Errorf("Failed to create a temp file for ca cert generation %v", err)
```
Comment (weekface) [inline tests/pkg/apimachinery/certs.go]: and all these should return a error.

File: tests/pkg/webhook/pods.go [hunk 4]
```diff
@@ -45,44 +72,95 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
-		content, err := ioutil.ReadAll(res.Body)
+		if info.IsOwner && tc.Status.TiDB.StatefulSet.Replicas > 1 {
+			time.Sleep(10 * time.Second)
+			glog.Errorf("tidb is ddl owner, can't be deleted namespace %s name %s", nameSpace, name)
+			os.Exit(3)
+		} else {
+			glog.Infof("savely delete pod namespace %s name %s content %s", nameSpace, name, string(content))
+		}
+
+	} else if pod.Labels["app.kubernetes.io/component"] == "pd" {
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]:
```suggestion
	} else if pod.Labels[label.ComponentLabelKey] == "pd" {
```

File: tests/pkg/webhook/pods.go [hunk 5]
```diff
@@ -45,44 +72,95 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
+			time.Sleep(10 * time.Second)
+			glog.Errorf("pd is leader, can't be deleted namespace %s name %s", nameSpace, name)
+			os.Exit(3)
+		} else {
+			glog.Infof("savely delete pod namespace %s name %s leader name %s", nameSpace, name, leader.Name)
 		}
 
-		if info.IsOwner {
-			glog.Errorf("tidb is ddl owner, can't be deleted namespace %s name %s", nameSpace, name)
+	} else if pod.Labels["app.kubernetes.io/component"] == "tikv" {
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 6]
```diff
@@ -45,44 +72,95 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
+		var storeID string
+		podIP := tc.Status.PD.Leader.ClientURL
+		for _, store := range tc.Status.TiKV.Stores {
+			if store.PodName == name {
+				storeID = store.ID
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]:
```suggestion
				storeID = store.ID
				break
```

File: tests/pkg/apimachinery/certs.go (hunk 1 above)
Comment (shuijing198799) [reply tests/pkg/apimachinery/certs.go]: all done

File: tests/cmd/e2e/main.go [hunk 7]
```diff
@@ -40,7 +40,10 @@ func main() {
 
 	cli, kubeCli := client.NewCliOrDie()
 
-	context := apimachinery.SetupServerCert(os.Getenv("NAMESPACE"), "webhook-service")
+	context, err := apimachinery.SetupServerCert(os.Getenv("NAMESPACE"), "webhook-service")
```
Comment (weekface) [inline tests/cmd/e2e/main.go]: merge this function into `StartValidatingAdmissionWebhookServerOrDie`.

File: tests/pkg/apimachinery/certs.go [hunk 8]
```diff
@@ -16,30 +16,35 @@ type CertContext struct {
...
 		glog.Errorf("Failed to create CA cert for apiserver %v", err)
+		return nil, err
 	}
 	caCertFile, err := ioutil.TempFile(certDir, "ca.crt")
 	if err != nil {
 		glog.Errorf("Failed to create a temp file for ca cert generation %v", err)
+		return nil, err
 	}
 	if err := ioutil.WriteFile(caCertFile.Name(), cert.EncodeCertPEM(signingCert), 0644); err != nil {
 		glog.Errorf("Failed to write CA cert %v", err)
```
Comment (weekface) [inline tests/pkg/apimachinery/certs.go]: return error here.

File: tests/cmd/e2e/main.go (hunk 7 above)
Comment (shuijing198799) [reply tests/cmd/e2e/main.go]: done

File: tests/pkg/apimachinery/certs.go (hunk 8 above)
Comment (weekface) [reply tests/pkg/apimachinery/certs.go]: this should be fixed also

Comment (shuijing198799) [reply tests/pkg/apimachinery/certs.go]: done

File: tests/pkg/webhook/pods.go [hunk 9]
```diff
@@ -156,14 +204,27 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
 			os.Exit(3)
 		} else {
 			glog.Infof("savely delete pod namespace %s name %s", nameSpace, name)
 		}
+
+		kvShoot[shootName] = true
+
+	}
+
+	if kvLeaders, err = getAllKVLeaders(tc, httpClient); err != nil {
```
Comment (xiaojingchen) [inline tests/pkg/webhook/pods.go]: the `getAllKVLeaders` should be called outside admitPods, best in the init phase of the webhook server.

File: tests/pkg/webhook/pods.go [hunk 10]
```diff
@@ -46,10 +52,42 @@ func HttpHandler(url string, method string, httpClient *http.Client) (content []
 
 }
 
+func getAllKVLeaders(tc *v1alpha1.TidbCluster, httpClient *http.Client) (ret map[string]int, err error) {
+	ret = make(map[string]int)
+	podIP := tc.Status.PD.Leader.ClientURL
+	for _, store := range tc.Status.TiKV.Stores {
+		url := fmt.Sprintf("%s/pd/api/v1/store/%s", podIP, store.ID)
```
[the remark this answers is not in the document]
Comment (shuijing198799) [reply tests/pkg/webhook/pods.go]: get store have no pod name。

[the remark this answers is not in the document]
Comment (weekface) [reply tests/pkg/webhook/pods.go]: what does it mean?

[the remark this answers is not in the document]
Comment (xiaojingchen) [reply tests/pkg/webhook/pods.go]: use `GetStore` is better,it only needs `storeID` and less codes

File: pkg/controller/tidb_control.go [hunk 11]
```diff
@@ -89,6 +96,37 @@ func (tdc *defaultTiDBControl) ResignDDLOwner(tc *v1alpha1.TidbCluster, ordinal
 	return false, err2
 }
 
+func (tdc *defaultTiDBControl) GetInfo(tc *v1alpha1.TidbCluster, ordinal int32) (*dbInfo, error) {
```
Comment (weekface) [inline pkg/controller/tidb_control.go]: Open an issue to remember to add unit tests to cover this method.

File: tests/pkg/webhook/pods.go [hunk 12]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
-	shootName := fmt.Sprintf("%s:%s", nameSpace, name)
+	namespace := ar.Request.Namespace
 
 	reviewResponse := v1beta1.AdmissionResponse{}
 	reviewResponse.Allowed = true
 
-	pod, err := kubeCli.CoreV1().Pods(nameSpace).Get(name, metav1.GetOptions{})
+	pod, err := kubeCli.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
 	if err != nil {
-		reviewResponse.Allowed = false
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: Why delete this line? `reviewResponse.Allowed` default is true.

File: tests/pkg/webhook/pods.go [hunk 13]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
+		glog.Infof("api server send wrong pod info namespace %s name %s err %v", namespace, name, err)
 		return &reviewResponse
 	}
 
 	glog.Infof("delete pod %s", pod.Labels[label.ComponentLabelKey])
 
-	tc, err := versionCli.PingcapV1alpha1().TidbClusters(nameSpace).Get(pod.Labels["app.kubernetes.io/instance"], metav1.GetOptions{})
+	tc, err := versionCli.PingcapV1alpha1().TidbClusters(namespace).Get(pod.Labels[label.InstanceLabelKey], metav1.GetOptions{})
 	if err != nil {
-		reviewResponse.Allowed = false
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 14]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
 	if pod.Labels[label.ComponentLabelKey] == "tidb" {
-		podIP := pod.Status.PodIP
-		url := fmt.Sprintf("http://%s:10080/info", podIP)
 
-		content, err := HttpHandler(url, "POST", httpClient)
+		ordinal, err := strconv.ParseInt(strings.Split(name, "-")[len(strings.Split(name, "-"))-1], 10, 32)
 		if err != nil {
-			glog.Errorf("fail to read response %v", err)
+			glog.Errorf("fail to convert string to int while deleting TiDB err %v", err)
 			return &reviewResponse
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 15]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
 		}
 
-		info := dbInfo{}
-		err = json.Unmarshal(content, &info)
+		info, err := tidbController.GetInfo(tc, int32(ordinal))
 		if err != nil {
-			glog.Errorf("unmarshal failed,namespace %s name %s error:%v", nameSpace, name, err)
+			glog.Errorf("fail to get tidb info error:%v", err)
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 16]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
-
 	} else if pod.Labels[label.ComponentLabelKey] == "pd" {
-		podIP := tc.Status.PD.Leader.ClientURL
-		url := fmt.Sprintf("%s/pd/api/v1/leader", podIP)
 
-		content, err := HttpHandler(url, "GET", httpClient)
+		leader, err := pdClient.GetPDLeader()
 		if err != nil {
-			glog.Errorf("fail to read response %v", err)
-			return &reviewResponse
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 17]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
-		}
-
-		leader := &pdpb.Member{}
-		err = json.Unmarshal(content, leader)
-		if err != nil {
-			glog.Errorf("unmarshal failed,namespace %s name %s error:%v", nameSpace, name, err)
+			glog.Errorf("fail to get pd leader %v", err)
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 18]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
-		podIP := tc.Status.PD.Leader.ClientURL
+		var storeID uint64
+		storeID = 0
 		for _, store := range tc.Status.TiKV.Stores {
 			if store.PodName == name {
-				storeID = store.ID
+				storeID, err = strconv.ParseUint(store.ID, 10, 64)
+				if err != nil {
+					glog.Errorf("fail to convert string to int while deleting PD err %v", err)
+					return &reviewResponse
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 19]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
 		}
 
-		url := fmt.Sprintf("%s/pd/api/v1/store/%s", podIP, storeID)
-
-		content, err := HttpHandler(url, "GET", httpClient)
-		if err != nil {
-			glog.Errorf("fail to read response %v", err)
+		// Fail to get store in stores
+		if storeID == 0 {
+			glog.Errorf("fail to find store in TIKV.Stores podname %s", name)
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

File: tests/pkg/webhook/pods.go [hunk 20]
```diff
@@ -99,132 +73,104 @@ func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
...
 			return &reviewResponse
 		}
 
-		storeInfo := &controller.StoreInfo{}
-		err = json.Unmarshal(content, storeInfo)
+		storeInfo, err := pdClient.GetStore(storeID)
 		if err != nil {
-			glog.Errorf("unmarshal failed,namespace %s name %s error:%v", nameSpace, name, err)
+			glog.Errorf("fail to read storeID %d response %v", storeID, err)
```
Comment (weekface) [inline tests/pkg/webhook/pods.go]: ditto

**Pull request merged**